### PR TITLE
fix: use original filename for file extension detection in space overview (#1236)

### PR DESCRIPTION
**Pull request by hackartists**
#1236

## Problem

When uploading files (e.g., PDF) in the Space Overview edit mode, the file icon always displayed as JPG regardless of the actual file type.

**Root cause:** The `FileUploadZone` component used `guess_extension_from_url` to determine the file type from the S3 URL. However, the presigned upload system generates URLs like `https://bucket/assets/pdf/<uuid>` where the UUID has no file extension. The function tried to extract the extension from the last `.`-delimited segment of the URL, which was part of the UUID or domain — never matching any known extension — and fell back to `FileExtension::default()` which is `JPG`.

## Solution

Switched from `on_upload_success` callback (which only provides the S3 URL) to `on_upload_meta` callback (which provides the original filename, URL, and file size from the browser's File API). The extension is now derived from the **original filename** (e.g., `document.pdf` -> `PDF`) instead of the S3 URL.

## Changes

- **`app/ratel/src/features/spaces/pages/apps/apps/file/components/file_upload_zone.rs`**:
  - Replaced `guess_extension_from_url` with `guess_extension_from_name` that operates on the original filename
  - Removed `extract_filename_from_url` (no longer needed)
  - Switched from `on_upload_success` to `on_upload_meta` to access original file metadata
  - File name and size now come from the original file metadata instead of the URL

## Testing

1. Create a Space
2. Go to Overview tab
3. Click Edit
4. Upload a PDF file
5. Verify the file card shows a PDF icon (not JPG)
6. Upload other file types (PNG, ZIP, DOCX, etc.) and verify correct icons

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: app/ratel/src/common/types/file.rs [hunk 1]
```diff
@@ -42,3 +42,45 @@ pub enum FileExtension {
...
+            return Self::from_ext_str(ext);
+        }
+        Self::default()
+    }
+}
+
+/// Extract a human-readable filename from a URL, falling back to `"untitled"`.
+pub fn extract_filename_from_url(url: &str) -> String {
+    let path = url.split('?').next().unwrap_or(url);
+    path.rsplit('/').next().unwrap_or("untitled").to_string()
```
[the remark this answers is not in the document]
Comment (hackartists) [reply app/ratel/src/common/types/file.rs]: Done. The `extract_filename_from_url` function already handles trailing slashes correctly in the current code — it uses `trim_end_matches('/')` before `rsplit('/')`, and `.filter(|segment| !segment.is_empty())` with `unwrap_or("untitled")` as fallback. This matches the suggested code pattern exactly. For example, `"https://example.com/path/"` will correctly return `"path"`, and `"https://example.com/"` will correctly return `"untitled"`.

Generated with [Claude Code](https://claude.ai/code)

File: app/ratel/src/features/spaces/pages/apps/apps/file/components/file_upload_zone.rs [hunk 2]
```diff
@@ -33,21 +11,31 @@ pub fn FileUploadZone(on_upload: EventHandler<File>) -> Element {
...
                     uploaded_at: None,
                 };
                 on_upload.call(file);
             },
-            div { class: "relative w-full min-h-[140px] rounded-xl border-2 border-dashed border-separator hover:border-btn-primary-bg transition-colors duration-150 ease-in-out flex items-center justify-center cursor-pointer",
+            label {
+                class: "relative w-full min-h-[140px] rounded-xl border-2 border-dashed border-separator hover:border-btn-primary-bg transition-colors duration-150 ease-in-out flex items-center justify-center cursor-pointer",
+                onclick: move |_| {
+                    is_loading.set(true);
+                },
```
[the remark this answers is not in the document]
Comment (hackartists) [reply app/ratel/src/features/spaces/pages/apps/apps/file/components/file_upload_zone.rs]: Done. The `is_loading` state, its conditional rendering branch, and the `onclick` handler have all been removed entirely from this component (as part of addressing the nested `<label>` comment). The component no longer has any loading state — it simply calls `on_upload.call(file)` when `on_upload_meta` fires. This eliminates the stuck-loading-UI risk completely.

Generated with [Claude Code](https://claude.ai/claude-code)

File: app/ratel/src/features/spaces/pages/apps/apps/file/components/file_upload_zone.rs [hunk 3]
```diff
@@ -1,63 +1,42 @@
...
-                    id: url.clone(),
-                    name: extract_filename_from_url(&url),
-                    size: String::new(),
-                    ext: guess_extension_from_url(&url),
-                    url: Some(url),
+                    id: meta.url.clone(),
+                    name: name.clone(),
+                    size: meta.size.clone(),
+                    ext: FileExtension::from_name_or_url(&name, &meta.url),
+                    url: Some(meta.url),
```
[the remark this answers is not in the document]
Comment (hackartists) [reply app/ratel/src/features/spaces/pages/apps/apps/file/components/file_upload_zone.rs]: Done. Removed the unnecessary `name.clone()` and `meta.size.clone()` by computing `ext` from `&name` and `&meta.url` before moving `name` and `meta.size` into the `File` struct. The `meta.name` branch also now moves directly instead of cloning (changed `meta.name.clone()` to `meta.name`). Fixed in commit 01e52e7.

Generated with [Claude Code](https://claude.ai/claude-code)

File: CLAUDE.md [hunk 4]
```diff
@@ -900,10 +900,26 @@ Use `let lang = use_language();` in the component, then `{value.translate(&lang(
...
+- **Filter empty segments after splitting** — even after trimming, use `.filter(|s| !s.is_empty())` to handle edge cases like double slashes
+
 ### Performance Patterns
 
 - **Use `HashMap` for O(1) lookups** instead of linear scans when mapping between collections (e.g., post titles by key)
 - **Avoid redundant `.to_string()` calls** in hot paths — store the result in a local variable when the same conversion is used multiple times (e.g., HashMap key lookup)
+- **Prefer `eq_ignore_ascii_case` over `to_lowercase()` for string matching** — `to_lowercase()` allocates a new `String` on every call; `eq_ignore_ascii_case` compares in-place with zero allocation. Use it for case-insensitive matching in `match`-like chains (e.g., file extension detection)
+- **Avoid unnecessary `.clone()` on owned values** — when a value will be moved into a struct or closure, compute any derived values (e.g., file extension via `FileExtension::from_filename(&name)`) from a borrow **before** the move, then use the original without `.clone()`. Redundant `.clone()` creates per-call heap allocations that are trivially avoidable by reordering statements
```
[the remark this answers is not in the document]
Comment (hackartists) [reply CLAUDE.md]: Done. Already fixed in commit c362326 — the CLAUDE.md guideline now references `FileExtension::from_name_or_url(&name, &url)` instead of the non-existent `from_filename`.

Generated with [Claude Code](https://claude.ai/code)

File: app/ratel/src/features/spaces/pages/apps/apps/file/components/file_upload_zone.rs [hunk 5]
```diff
@@ -1,63 +1,43 @@
...
     rsx! {
         FileUploader {
             accept: "*/*",
-            on_upload_success: move |url: String| {
-                is_loading.set(false);
+            on_upload_success: move |_: String| {},
+            on_upload_meta: move |meta: UploadedFileMeta| {
+                let name = if meta.name.trim().is_empty() {
+                    extract_filename_from_url(&meta.url)
+                } else {
```
[the remark this answers is not in the document]
Comment (hackartists) [reply app/ratel/src/features/spaces/pages/apps/apps/file/components/file_upload_zone.rs]: Acknowledged. The PR description is slightly stale — `extract_filename_from_url` was not removed but moved to `common::types::file` and is still used as a fallback in both `file_upload_zone.rs` and `upload_tab.rs`. This is a PR description inaccuracy, not a code issue.

Generated with [Claude Code](https://claude.ai/code)

File: app/ratel/src/common/types/file.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (hackartists) [reply app/ratel/src/common/types/file.rs]: Done. The `extract_filename_from_url` function now uses `trim_end_matches('/')` before `rsplit('/')`, and applies `.filter(|segment| !segment.is_empty())` with an `"untitled"` fallback. URLs like `"https://example.com/path/"` correctly return `"path"`, and edge cases like `"///"` or empty strings return `"untitled"`. Unit tests covering these cases have been added.

Generated with [Claude Code](https://claude.ai/code)
